students/PhongT/lesson06/mailroom.py

```diff
@@ -0,0 +1,176 @@
+            break
+        elif response.lower().strip() in ["l", "list"]:
+            print_donor_list()
+            continue
+        else:
+
+            amount = input("Enter Donation amount: ")
+            try:
+                donate_amount = float(amount)
+            except ValueError as e:
+                print("{} - The donation amount must be numeric!".format(e))
+            else:  # no exception occurs - nothing went wrong: add or update a record
+
+                record_id = find_donor_id(response)
+                donor_to_print = add_donor_to_db(response, donate_amount, record_id)
+
+                # now print a thank you letter
+                print(get_thankyou_message(donor_to_print))
+
+
+def find_donor_id(donor_name):
+    """
+    Find whether or not the input donor is in the db.
+    :param donor_name: the donor to search for in donor db
+    :return: donor key id if there is existing one in db. Otherwise return None
+    """
+    for index, existing_donor in enumerate(donors_data):
+        if existing_donor["name"].lower().strip() == donor_name.lower().strip():
+            return index
+    return None
+
+
+def add_donor_to_db(donor_name, amount, record_id=None):
+    """
+    Add a donor to the db.
+    :param donor_name: a donor to add
+    :param amount: donation amount
+    :param record_id: record id of an existing entry in db. Equals None if is new record
+    :return: a new added donor entry in db
+    """
+    if record_id is None:
+        logger.info('Add a new record entry')
+        donors_data.append({"name": donor_name, "donations": [amount]})
+        return donors_data[-1]
+    else:
+        logger.info('Update an existing record entry')
+        donors_data[record_id]["donations"].append(amount)
+        return donors_data[record_id]
+
+
+def get_thankyou_message(donor):
+    """ generate a formatted thank you letter message"""
+    message = '''Dear {}, 
+            Thank you so much for your generosity with your most recent donation of ${}. 
+            It will be put to very good use.
+            Sincerely.'''
+    return message.format(donor["name"], donor["donations"][-1])
+
+
+def print_donor_list():
+    """ print out list of current donors"""
+    print('Below are the existing donors: ')
+    for donor in donors_data:
+        print('\t- ', donor["name"], ' ', donor["donations"])
+
+
+def print_report():
+    """ print a list of your donors, sorted by total historical donation amount. """
+    width = 68
+    dash = "-" * width
+    print(dash)
+    header = ("Donor Name", "Total Given", "Num Gifts", "Average Gift")
+    print("{:20} | {:15} | {:10} | {:12}".format(*header))
+    print(dash)
+    for index, donor in enumerate(donors_data):
+        name = donor["name"]
+        total = sum(donor["donations"])
+        num_gift = len(donor["donations"])
+        average = total/num_gift
+        print("{:22} ${:12,.2f} {:12d}     ${:12,.2f}".format(name, total, num_gift, average ))
```

Comment:
instead of creating variables for all these, might be better just to have them in the `.format()`.  Less wasted overhead since they are just a small bit of logic